File: src/script/libs/Assets.zig
// Copyright 2026 wyteroze. Licensed under the Apache License, Version 2.0.

const std = @import("std");
const MeshData = @import("../../MeshData.zig").MeshData;
const ImageData = @import("../../ImageData.zig").ImageData;

const mesh_path = "src/assets/models/";
const image_path = "src/assets/images/";

pub const Assets = struct {
    allocator: std.mem.Allocator,
    io: std.Io,

    pub fn init(allocator: std.mem.Allocator, io: std.Io) Assets {
        return .{
            .allocator = allocator,
            .io = io
        };
    }

    pub fn loadMesh(self: *Assets, path: []const u8) !MeshData {
        const full_path = try std.mem.concat(self.allocator, u8, &.{ mesh_path, path });
        defer self.allocator.free(full_path);

        return try MeshData.loadFromFile(self.allocator, self.io, full_path);
    }

    pub fn loadImage(self: *Assets, path: []const u8) !ImageData {
        const full_path = try std.mem.concat(self.allocator, u8, &.{ image_path, path });
        defer self.allocator.free(full_path);

        return try ImageData.loadFromFile(self.allocator, self.io, full_path);
    }
};

File: scripts/api/Assets.lua
--[[
Copyright 2026 wyteroze. Licensed under the Apache License, Version 2.0.
--]]

--- @meta Assets
-- This file is for the Lua Language Server, do not require it

--- Factory for loading assets
--- @class AssetsLib
Assets = {}

--- Loads a mesh from your game's `assets/models/`
--- You may also choose meshes from folders. (ex. `guns/pistol/bullet.obj`)
--- @param meshPath string
--- @return MeshData
function Assets.loadMesh(meshPath) end

--- Loads an image from your game's `assets/images/`
--- You may also choose images from folders. (ex. `guns/pistol/icon.bmp`)
--- @param imagePath string
--- @return ImageData
function Assets.loadImage(imagePath) end

--- Loads an audio from your game's `assets/audios/`
--- You may also choose audios from folders. (ex. `guns/pistol/shoot.wav`)
--- @param audioPath string
--- @return AudioData
function Assets.loadAudio(audioPath) end

--- Loads a font from your game's `assets/fonts/`
--- Expects a folder containing `map.toml` and its character sheet. (ex. `default`)
--- @param fontName string
--- @return Font
function Assets.loadFont(fontName) end
